File: fastbloom_rs/filter.py
# -*- coding: utf-8 -*-

from typing import Union, Sequence

from fastbloom_rs import PyFilterBuilder, PyBloomFilter, PyCountingBloomFilter


class FilterBuilder(object):
    """
    Constructs a new Bloom Filter Builder by specifying the expected size of the filter and the
    tolerable false positive probability. The size of the BLoom filter in in bits and the
    optimal number of hash functions will be inferred from this.

    :param expected_elements: expected size of the filter
    :param false_positive_probability: tolerable false positive probability
    """

    def __init__(self, expected_elements: Union[int, PyFilterBuilder], false_positive_probability: float = 0.01):
        if isinstance(expected_elements, int):
            self._py_builder = PyFilterBuilder(expected_elements, false_positive_probability)
        elif isinstance(expected_elements, PyFilterBuilder):
            self._py_builder = expected_elements
        else:
            raise Exception("expected_elements must be integer")

    def expected_elements(self) -> int:
        """
        expected size of the filter

        :return:
        """
        return self._py_builder.expected_elements()

    def false_positive_probability(self) -> float:
        """
        tolerable false positive probability

        :return:
        """
        return self._py_builder.false_positive_probability()

    def enable_repeat_insert(self, enable: bool):
        """
        Whether to allow multiple inserts of the same element. Only use for Counting Bloom Filter.

        :param enable:
        :return:
        """
        self._py_builder.enable_repeat_insert(enable)

    def size(self) -> int:
        """
        the size of the bloom filter in bits.

        :return:
        """
        return self._py_builder.size()

    def hashes(self) -> int:
        """
         the hash function number of the Bloom filter.

        :return:
        """
        return self._py_builder.hashes()

    def build_bloom_filter(self) -> "BloomFilter":
        """
        Constructs a Bloom filter using the specified parameters and computing missing parameters
        if possible (e.g. the optimal Bloom filter bit size).

        :return:
        """
        return BloomFilter(self._py_builder.build_bloom_filter())

    def build_counting_bloom_filter(self) -> "CountingBloomFilter":
        """
        Constructs a Counting Bloom filter using the specified parameters and computing missing parameters
        if possible (e.g. the optimal Bloom filter counter size).

        :return:
        """
        return CountingBloomFilter(self._py_builder.build_counting_bloom_filter())


class BloomFilter(object):
    """
    A bloom filter powered by rust.

    :param expected_elements: expected elements in the filter
    :param false_positive_probability: tolerable false positive probability
    """

    def __init__(self, expected_elements: Union[int, PyBloomFilter], false_positive_probability: float = 0.01):
        if isinstance(expected_elements, int):
            self._py_builder = PyFilterBuilder(expected_elements, false_positive_probability)
            self._py_bloom = self._py_builder.build_bloom_filter()
        elif isinstance(expected_elements, PyBloomFilter):
            self._py_bloom = expected_elements
        else:
            raise Exception("expected_elements must be integer")

    def add(self, element: Union[str, int, bytes]):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        if isinstance(element, int):
            self._py_bloom.add_int(element)
        elif isinstance(element, str):
            self._py_bloom.add_str(element)
        elif isinstance(element, bytes):
            self._py_bloom.add_bytes(element)
        else:
            self._py_bloom.add_str(str(element))

    def add_if_not_contains(self, element: Union[str, int, bytes]) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).
        And if it is not in this filter, add it to the filter.

        :param element: value to test
        :return: “False” if this element did not exist in the Bloom filter before, and then this method will insert
        this element into the current filter. “True” if the element is already in the Bloom filter.
        """
        if isinstance(element, int):
            return self._py_bloom.add_int_if_not_contains(element)
        elif isinstance(element, str):
            return self._py_bloom.add_str_if_not_contains(element)
        elif isinstance(element, bytes):
            return self._py_bloom.add_bytes_if_not_contains(element)
        else:
            return self._py_bloom.add_str_if_not_contains(str(element))

    def add_int(self, element: int):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        self._py_bloom.add_int(element)

    def add_int_batch(self, array: Sequence[int]):
        self._py_bloom.add_int_batch(array)

    def add_str(self, element: str):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        self._py_bloom.add_str(element)

    def add_str_batch(self, array: Sequence[str]):
        self._py_bloom.add_str_batch(array)

    def add_bytes(self, element: bytes):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        self._py_bloom.add_bytes(element)

    def add_bytes_batch(self, elements: Sequence[bytes]):
        """
        Add all bytes to the filter.
        :param elements: byte list
        :return:
        """
        self._py_bloom.add_bytes_batch(elements)

    def contains(self, element: Union[str, int, bytes]) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        if isinstance(element, int):
            return self._py_bloom.contains_int(element)
        elif isinstance(element, str):
            return self._py_bloom.contains_str(element)
        elif isinstance(element, bytes):
            return self._py_bloom.contains_bytes(element)
        else:
            return self._py_bloom.contains_str(str(element))

    def contains_int(self, element: int) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        return self._py_bloom.contains_int(element)

    def contains_int_batch(self, elements: Sequence[int], check_type: bool = True) -> Sequence[bool]:
        """
        Check all int element whether in the filter.

        :param elements: elements to check
        :param check_type: whether to check elements in elements is type int
        :return: a bool sequence for each element in elements is whether in this filter.
        """
        if check_type:
            for ele in elements:
                assert isinstance(ele, int)
        return self._py_bloom.contains_int_batch(elements)

    def contains_str(self, element: str) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        return self._py_bloom.contains_str(element)

    def contains_str_batch(self, elements: Sequence[str], check_type: bool = True) -> Sequence[bool]:
        """
        Check all int element whether in the filter.

        :param elements: elements to check
        :param check_type: whether to check elements in elements is type str
        :return: a bool sequence for each element in elements is whether in this filter.
        """
        if check_type:
            for ele in elements:
                assert isinstance(ele, str)
        return self._py_bloom.contains_str_batch(elements)

    def contains_bytes(self, element: bytes) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        return self._py_bloom.contains_bytes(element)

    def contains_bytes_batch(self, elements: Sequence[bytes], check_type: bool = True) -> Sequence[bool]:
        """
        Check all int element whether in the filter.

        :param elements: elements to check
        :param check_type: whether to check elements in elements is type bytes
        :return: a bool sequence for each element in elements is whether in this filter.
        """
        if check_type:
            for ele in elements:
                assert isinstance(ele, bytes)
        return self._py_bloom.contains_bytes_batch(elements)

    def contains_hash_indices(self, indices: Sequence[int]) -> bool:
        """
        Tests whether a hashes indices is present in the filter
        :param indices: indices array
        :return:
        """
        return self._py_bloom.contains_hash_indices(indices)

    def get_hash_indices(self, element: Union[str, int, bytes]) -> Sequence[int]:
        """
        Get the hashes indices of the element in the filter.
        :param element: to compute
        :return: indices array
        """
        if isinstance(element, int):
            return self._py_bloom.get_hash_indices_int(element)
        elif isinstance(element, str):
            return self._py_bloom.get_hash_indices_str(element)
        elif isinstance(element, bytes):
            return self._py_bloom.get_hash_indices(element)
        else:
            return self._py_bloom.get_hash_indices_str(str(element))

    def config(self) -> FilterBuilder:
        """
        Returns the configuration/builder of the Bloom filter.

        :return:
        """
        return FilterBuilder(self._py_bloom.config())

    def hashes(self) -> int:
        """
        Returns the hash function number of the Bloom filter.

        :return:
        """
        return self._py_bloom.hashes()

    def get_bytes(self) -> bytes:
        """
        Return the underlying byte vector of the Bloom filter.

        :return:
        """
        return self._py_bloom.get_bytes()

    def get_int_array(self) -> Sequence[int]:
        """
        Return the underlying u32 vector of the Bloom filter.

        :return:
        """
        return self._py_bloom.get_int_array()

    def clear(self):
        """
        Removes all elements from the filter (i.e. resets all bits to zero).

        :return:
        """
        self._py_bloom.clear()

    def is_empty(self) -> bool:
        """
        Returns [true] if the Bloom filter does not contain any elements

        :return:
        """
        return self._py_bloom.is_empty()

    def union(self, other: "BloomFilter") -> bool:
        """
        Performs the union operation on two compatible bloom filters. This is achieved through a
        bitwise OR operation on their bit vectors. This operations is lossless, i.e. no elements
        are lost and the bloom filter is the same that would have resulted if all elements wer
        directly inserted in just one bloom filter.

        :param other:
        :return:
        """
        return self._py_bloom.union(other._py_bloom)

    def intersect(self, other: "BloomFilter") -> bool:
        """
        Performs the intersection operation on two compatible bloom filters. This is achieved
        through a bitwise AND operation on their bit vectors. The operations doesn't introduce
        any false negatives but it does raise the false positive probability. The the false
        positive probability in the resulting Bloom filter is at most the false-positive probability
        in one of the constituent bloom filters

        :param other:
        :return:
        """
        return self._py_bloom.intersect(other._py_bloom)

    def __contains__(self, item: Union[str, int, bytes]):
        return self.contains(item)

    @staticmethod
    def from_bytes(array: bytes, hashes: int) -> "BloomFilter":
        """
        Build a Bloom filter form [u8].

        :param array: byte array
        :param hashes: hash function number of the Bloom filter
        :return:
        """
        py_bloom = PyBloomFilter.from_bytes(array, hashes)
        return BloomFilter(py_bloom)

    @staticmethod
    def from_int_array(array: Sequence[int], hashes: int) -> "BloomFilter":
        """
        Build a Bloom filter form [u32].

        :param array: integer(u32) array
        :param hashes: hash function number of the Bloom filter
        :return:
        """
        py_bloom = PyBloomFilter.from_int_array(array, hashes)
        return BloomFilter(py_bloom)


class CountingBloomFilter(object):
    """
    A Counting Bloom Filter powered by rust. The counter is four bits.

    :param expected_elements: expected elements in the filter
    :param false_positive_probability: tolerable false positive probability
    :param enable_repeat_insert: Whether to allow multiple inserts of the same element.
    """

    def __init__(self, expected_elements: Union[int, PyCountingBloomFilter], false_positive_probability: float = 0.01,
                 enable_repeat_insert: bool = True):
        if isinstance(expected_elements, int):
            self._py_builder = PyFilterBuilder(expected_elements, false_positive_probability)
            self._py_builder.enable_repeat_insert(enable_repeat_insert)
            self._py_counting_bloom = self._py_builder.build_counting_bloom_filter()
        elif isinstance(expected_elements, PyCountingBloomFilter):
            self._py_counting_bloom = expected_elements
        else:
            raise Exception("expected_elements must be integer")

    def add(self, element: Union[str, int, bytes]):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        if isinstance(element, int):
            self._py_counting_bloom.add_int(element)
        elif isinstance(element, str):
            self._py_counting_bloom.add_str(element)
        elif isinstance(element, bytes):
            self._py_counting_bloom.add_bytes(element)
        else:
            self._py_counting_bloom.add_str(str(element))

    def remove(self, element: Union[str, int, bytes]):
        """
        Remove element to the filter.

        :param element:
        :return:
        """
        if isinstance(element, int):
            self._py_counting_bloom.remove_int(element)
        elif isinstance(element, str):
            self._py_counting_bloom.remove_str(element)
        elif isinstance(element, bytes):
            self._py_counting_bloom.remove_bytes(element)
        else:
            self._py_counting_bloom.remove_str(str(element))

    def add_int(self, element: int):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        self._py_counting_bloom.add_int(element)

    def add_int_batch(self, array: Sequence[int]):
        self._py_counting_bloom.add_int_batch(array)

    def remove_int(self, element: int):
        """
        Remove element from this filter.

        :param element:
        :return:
        """
        self._py_counting_bloom.remove_int(element)

    def add_str(self, element: str):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        self._py_counting_bloom.add_str(element)

    def add_str_batch(self, array: Sequence[str]):
        self._py_counting_bloom.add_str_batch(array)

    def remove_str(self, element: str):
        """
        Remove element from this filter.

        :param element:
        :return:
        """
        self._py_counting_bloom.remove_str(element)

    def add_bytes(self, element: bytes):
        """
        Add element to the filter.

        :param element: value to add
        :return:
        """
        self._py_counting_bloom.add_bytes(element)

    def add_bytes_batch(self, elements: Sequence[bytes]):
        """
        Add all bytes to the filter.

        :param elements: byte list
        :return:
        """
        self._py_counting_bloom.add_bytes_batch(elements)

    def remove_bytes(self, element: bytes):
        """
        Remove element from this filter.

        :param element:
        :return:
        """
        self._py_counting_bloom.remove_bytes(element)

    def contains(self, element: Union[str, int, bytes]) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        if isinstance(element, int):
            return self._py_counting_bloom.contains_int(element)
        elif isinstance(element, str):
            return self._py_counting_bloom.contains_str(element)
        elif isinstance(element, bytes):
            return self._py_counting_bloom.contains_bytes(element)
        else:
            return self._py_counting_bloom.contains_str(str(element))

    def contains_int(self, element: int) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        return self._py_counting_bloom.contains_int(element)

    def contains_int_batch(self, elements: Sequence[int], check_type: bool = True) -> Sequence[bool]:
        """
        Check all int element whether in the filter.

        :param elements: elements to check
        :param check_type: whether to check elements in elements is type int
        :return: a bool sequence for each element in elements is whether in this filter.
        """
        if check_type:
            for ele in elements:
                assert isinstance(ele, int)
        return self._py_counting_bloom.contains_int_batch(elements)

    def contains_str(self, element: str) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        return self._py_counting_bloom.contains_str(element)

    def contains_str_batch(self, elements: Sequence[str], check_type: bool = True) -> Sequence[bool]:
        """
        Check all int element whether in the filter.

        :param elements: elements to check
        :param check_type: whether to check elements in elements is type str
        :return: a bool sequence for each element in elements is whether in this filter.
        """
        if check_type:
            for ele in elements:
                assert isinstance(ele, str)
        return self._py_counting_bloom.contains_str_batch(elements)

    def contains_bytes(self, element: bytes) -> bool:
        """
        Tests whether an element is present in the filter (subject to the specified false positive rate).

        :param element: to test
        :return: bool
        """
        return self._py_counting_bloom.contains_bytes(element)

    def contains_bytes_batch(self, elements: Sequence[bytes], check_type: bool = True) -> Sequence[bool]:
        """
        Check all int element whether in the filter.

        :param elements: elements to check
        :param check_type: whether to check elements in elements is type bytes
        :return: a bool sequence for each element in elements is whether in this filter.
        """
        if check_type:
            for ele in elements:
                assert isinstance(ele, bytes)
        return self._py_counting_bloom.contains_bytes_batch(elements)

    def contains_hash_indices(self, indices: Sequence[int]) -> bool:
        """
        Tests whether a hashes indices is present in the filter
        :param indices: indices array
        :return:
        """
        return self._py_counting_bloom.contains_hash_indices(indices)

    def get_hash_indices(self, element: Union[str, int, bytes]) -> Sequence[int]:
        """
        Get the hashes indices of the element in the filter.
        :param element: to compute
        :return: indices array
        """
        if isinstance(element, int):
            return self._py_counting_bloom.get_hash_indices_int(element)
        elif isinstance(element, str):
            return self._py_counting_bloom.get_hash_indices_str(element)
        elif isinstance(element, bytes):
            return self._py_counting_bloom.get_hash_indices(element)
        else:
            return self._py_counting_bloom.get_hash_indices_str(str(element))

    def estimate_count(self, element: Union[str, int, bytes]) -> int:
        """
        Get the estimate count for element in this counting bloom filter.
        See: https://github.com/yankun1992/fastbloom/issues/3

        :param element:
        :return:
        """
        if isinstance(element, int):
            return self._py_counting_bloom.estimate_count_int(element)
        elif isinstance(element, str):
            return self._py_counting_bloom.estimate_count_str(element)
        elif isinstance(element, bytes):
            return self._py_counting_bloom.estimate_count(element)
        else:
            return self._py_counting_bloom.estimate_count_str(str(element))

    def counter_at(self, index: int) -> int:
        """
        Get the underlying counter at index.

        :param index: index of counter slot.
        :return:
        """
        assert index >= 0
        return self._py_counting_bloom.counter_at(index)

    def config(self) -> FilterBuilder:
        """
        Returns the configuration/builder of the Bloom filter.

        :return:
        """
        return FilterBuilder(self._py_counting_bloom.config())

    def hashes(self) -> int:
        """
        Returns the hash function number of the Bloom filter.

        :return:
        """
        return self._py_counting_bloom.hashes()

    def get_bytes(self) -> bytes:
        """
        Return the underlying byte vector of the Bloom filter.

        :return:
        """
        return self._py_counting_bloom.get_bytes()

    def get_int_array(self) -> Sequence[int]:
        """
        Return the underlying u32 vector of the Bloom filter.

        :return:
        """
        return self._py_counting_bloom.get_int_array()

    def clear(self):
        """
        Removes all elements from the filter (i.e. resets all bits to zero).

        :return:
        """
        self._py_counting_bloom.clear()

    def __contains__(self, item: Union[str, int, bytes]):
        return self.contains(item)

    @staticmethod
    def from_bytes(array: bytes, hashes: int, enable_repeat_insert: bool = True) -> "CountingBloomFilter":
        """
        Build a Counting Bloom filter form [u8].

        :param enable_repeat_insert:
        :param array: byte array
        :param hashes: hash function number of the Bloom filter
        :return:
        """
        py_bloom = PyCountingBloomFilter.from_bytes(array, hashes, enable_repeat_insert)
        return CountingBloomFilter(py_bloom)

    @staticmethod
    def from_int_array(array: Sequence[int], hashes: int, enable_repeat_insert: bool = True) -> "CountingBloomFilter":
        """
        Build a Counting Bloom filter form [u32].

        :param enable_repeat_insert:
        :param array: integer(u32) array
        :param hashes: hash function number of the Bloom filter
        :return:
        """
        py_bloom = PyCountingBloomFilter.from_int_array(array, hashes, enable_repeat_insert)
        return CountingBloomFilter(py_bloom)

File: fastbloom-rs/src/filter.rs
use std::mem;
use std::ops::Index;
use std::ptr::slice_from_raw_parts;

use bit_vec::BitVec;
use fastmurmur3::murmur3_x64_128;
use smallvec::SmallVec;

use crate::builder::FilterBuilder;

#[inline]
fn bit_set(bit_set: &mut BitVec, value: &[u8], m: u128, k: u64) {
    // let len = m >> 5;
    let hash1 = (murmur3_x64_128(value, 0) % m) as u64;
    let hash2 = (murmur3_x64_128(value, 32) % m) as u64;

    let m = m as u64;
    for i in 1..k {
        let mo = ((hash1 + i * hash2) % m) as usize;
        bit_set.set(mo, true);
    };
    bit_set.set(hash1 as usize, true);
}

#[inline]
fn bit_check(bit_set: &BitVec, value: &[u8], m: u128, k: u64) -> bool {
    let hash1 = (murmur3_x64_128(value, 0) % m) as u64;
    let hash2 = (murmur3_x64_128(value, 32) % m) as u64;
    let mut res = bit_set.get(hash1 as usize).unwrap();
    if !res { return false; }
    let m = m as u64;
    for i in 1..k {
        let mo = ((hash1 + i * hash2) % m) as usize;
        res = res && bit_set.get(mo).unwrap();
        if !res { return false; }
    }
    res
}

#[derive(Clone)]
#[derive(Debug)]
pub struct BloomFilter {
    config: FilterBuilder,
    bit_set: BitVec,
}

impl BloomFilter {
    /// Build a Bloom filter form [FilterBuilder].
    ///
    /// # Examples:
    ///
    /// ```rust
    /// use fastbloom_rs::{BloomFilter, FilterBuilder};
    ///
    /// let builder = FilterBuilder::new(100_000_000, 0.01);
    /// let bloom = BloomFilter::new(builder);
    /// ```
    pub fn new(mut config: FilterBuilder) -> Self {
        config.complete();
        let bit_set = BitVec::from_elem(config.size as usize, false);
        BloomFilter { config, bit_set }
    }

    /// Build a Bloom filter form [BitVec].
    ///
    /// # Examples
    ///
    /// ```rust
    /// use bit_vec::BitVec;
    /// use fastbloom_rs::BloomFilter;
    /// let bit_vec = BitVec::from_elem(32768, false);
    /// let bloom = BloomFilter::from_bit_vec(&bit_vec, 4);
    /// ```
    pub fn from_bit_vec(bit_vec: &BitVec, hashes: u32) -> Self {
        let mut config =
            FilterBuilder::from_size_and_hashes(bit_vec.len() as u64, hashes);
        config.complete();
        BloomFilter { config, bit_set: bit_vec.clone() }
    }

    /// Build a Bloom filter form [BitVec].
    ///
    /// # Examples
    ///
    /// ```rust
    /// use bit_vec::BitVec;
    /// use fastbloom_rs::BloomFilter;
    /// let mut array = Vec::with_capacity(4096);
    /// unsafe {array.set_len(4096); array.fill(0u8);};
    /// let bloom = BloomFilter::from_u8_array(array.as_bytes(), 4);
    /// ```
    pub fn from_u8_array(array: &[u8], hashes: u32) -> Self {
        let mut config =
            FilterBuilder::from_size_and_hashes((array.len() * 8) as u64, hashes);
        config.complete();
        let mut bit_vec = BitVec::from_elem(config.size as usize, false);
        let bit_underlying = unsafe { bit_vec.storage_mut() };
        let u8_ptr = array.as_ptr() as *const u32;
        let u32_array = slice_from_raw_parts(u8_ptr, (config.size >> 5) as usize);
        bit_underlying.copy_from_slice(unsafe { &*u32_array });

        BloomFilter { config, bit_set: bit_vec }
    }

    /// Build a Bloom filter form [&\[u32\]].
    ///
    /// # Examples
    ///
    /// ```rust
    /// use bit_vec::BitVec;
    /// use fastbloom_rs::BloomFilter;
    /// let mut array = Vec::with_capacity(4096);
    /// unsafe {array.set_len(4096); array.fill(0u32);};
    /// let bloom = BloomFilter::from_u32_array(array.as_slice(), 4);
    /// ```
    pub fn from_u32_array(array: &[u32], hashes: u32) -> Self {
        let mut config =
            FilterBuilder::from_size_and_hashes((array.len() * 32) as u64, hashes);
        config.complete();
        let mut bit_vec = BitVec::from_elem(array.len() * 32, false);
        let bit_underlying = unsafe { bit_vec.storage_mut() };
        bit_underlying.copy_from_slice(array);
        // for (b, a) in bit_underlying.iter_mut().zip(array.iter()) {
        //     *b = *a;
        // }
        BloomFilter { config, bit_set: bit_vec }
    }

    /// Returns the configuration/builder of the Bloom filter.
    /// # Examples
    ///
    /// ```rust
    /// use fastbloom_rs::{BloomFilter, FilterBuilder};
    ///
    /// let bloom = FilterBuilder::new(100_000_000, 0.01).build_bloom_filter();
    /// let builder = bloom.config();
    /// ```
    ///
    pub fn config(&self) -> FilterBuilder {
        self.config.clone()
    }

    ///  Returns the hash function number of the Bloom filter.
    pub fn hashes(&self) -> u32 {
        self.config.hashes
    }

    /// Adds the passed value to the filter.
    pub fn add(&mut self, element: &[u8]) {
        bit_set(&mut self.bit_set, element, self.config.size as u128,
                self.config.hashes as u64);
    }

    /// Removes all elements from the filter (i.e. resets all bits to zero).
    pub fn clear(&mut self) {
        self.bit_set.clear();
    }

    /// Tests whether an element is present in the filter (subject to the specified false
    /// positive rate).
    #[inline]
    pub fn contains(&self, element: &[u8]) -> bool {
        bit_check(&self.bit_set, element, self.config.size as u128,
                  self.config.hashes as u64)
    }

    pub fn get(&self, index: usize) -> Option<bool> {
        self.bit_set.get(index)
    }

    pub fn set(&mut self, index: usize, to: bool) {
        self.bit_set.set(index, to)
    }

    /// Return the underlying bit vector of the Bloom filter.
    pub fn get_bit_vec(&self) -> BitVec {
        self.bit_set.clone()
    }

    /// Return the underlying byte vector of the Bloom filter.
    pub fn get_u8_array(&self) -> &[u8] {
        let storage = self.bit_set.storage();
        let u32_ptr = storage.as_ptr();
        let u8_ptr = u32_ptr as *const u8;
        let ptr = slice_from_raw_parts(u8_ptr, storage.len() * 4);
        unsafe { &*ptr }
    }

    /// Return the underlying u32 vector of the Bloom filter.
    pub fn get_u32_array(&self) -> &[u32] {
        self.bit_set.storage()
    }

    /// Performs the union operation on two compatible bloom filters. This is achieved through a
    /// bitwise OR operation on their bit vectors. This operations is lossless, i.e. no elements
    /// are lost and the bloom filter is the same that would have resulted if all elements wer
    /// directly inserted in just one bloom filter.
    pub fn union(&mut self, other: &BloomFilter) -> bool {
        if self.compatible(other) {
            self.bit_set.or(&other.bit_set);
            true
        } else { false }
    }

    /// Performs the intersection operation on two compatible bloom filters. This is achieved
    /// through a bitwise AND operation on their bit vectors. The operations doesn't introduce
    /// any false negatives but it does raise the false positive probability. The the false
    /// positive probability in the resulting Bloom filter is at most the false-positive probability
    /// in one of the constituent bloom filters
    pub fn intersect(&mut self, other: &BloomFilter) -> bool {
        if self.compatible(other) {
            self.bit_set.and(&other.bit_set);
            true
        } else { false }
    }

    /// Returns [true] if the Bloom filter does not contain any elements
    pub fn is_empty(&self) -> bool {
        self.bit_set.is_empty()
    }

    pub fn set_bit_vec(&mut self, bit_vec: BitVec) {
        assert_eq!(self.config.size, bit_vec.capacity() as u64);
        self.bit_set = bit_vec
    }

    /// Checks if two Bloom filters are compatible, i.e. have compatible parameters (hash function,
    /// size, etc.)
    fn compatible(&self, other: &BloomFilter) -> bool {
        self.config.is_compatible_to(&other.config)
    }
}

#[test]
fn shift_test() {
    assert_eq!(32 >> 5, 1);
}

#[test]
fn bloom_test() {
    let mut builder =
        FilterBuilder::new(10_000, 0.01);
    let mut bloom = builder.build_bloom_filter();
    println!("{:?}", bloom.config);
    bloom.add(b"hello");
    assert_eq!(bloom.contains(b"hello"), true);
    assert_eq!(bloom.contains(b"world"), false);

    let storage = &bloom.bit_set.storage()[0..300];
    println!("{:?}", storage);

    let mut bloom2 =
        BloomFilter::from_bit_vec(&bloom.bit_set, bloom.config.hashes);
    assert_eq!(bloom2.compatible(&bloom), true);
    assert_eq!(bloom2.contains(b"hello"), true);
    assert_eq!(bloom2.contains(b"world"), false);

    let mut bloom3 =
        BloomFilter::from_u32_array(bloom.get_u32_array(), bloom.config.hashes);
    assert_eq!(bloom3.compatible(&bloom), true);
    assert_eq!(bloom3.contains(b"hello"), true);
    assert_eq!(bloom3.contains(b"world"), false);

    let u8_array = bloom.get_u8_array();
    let mut bloom4 = BloomFilter::from_u8_array(u8_array, bloom.config.hashes);
    println!("{:?}", &bloom4.bit_set.storage()[0..300]);
    assert_eq!(bloom4.compatible(&bloom), true);
    assert_eq!(bloom4.contains(b"hello"), true);
    assert_eq!(bloom4.contains(b"world"), false);

    bloom2.add(b"hello world");

    assert_eq!(bloom.intersect(&bloom2), true);
    assert_eq!(bloom.contains(b"hello"), true);
    assert_eq!(bloom.contains(b"hello world"), false);

    bloom3.add(b"hello world");
    bloom3.add(b"hello yankun");

    assert_eq!(bloom3.union(&bloom2), true);
    assert_eq!(bloom3.contains(b"hello"), true);
    assert_eq!(bloom3.contains(b"hello world"), true);
    assert_eq!(bloom3.contains(b"hello yankun"), true);
}